Update plot_network to use the geopandas plotting API in place of the networkx plotting API

kbonney opened the pull request:
## Summary
This PR updates the `plot_network` function by replacing `networkx` plotting functions with `geopandas` plotting functions. This builds on the geospatial capabilities of `wntr` by using the `GeoDataFrames` generated by `to_gis()` to interface with the `geopandas` plotting capabilities. This update reduces the complexity of the plotting code, making it easier to maintain and opening opportunities for other plotting options.

These updates aim to minimize changes to the output of `plot_network`, except where changes add functionality.
The arguments of `plot_network` remain unchanged, with the exception of some additional keyword arguments with defaults, so these updates should be fully backwards compatible. 

New functionality includes:
- Vertices of links are incorporated in the plot, rather than always having straight lines between nodes.
- Valves, pumps, tanks and reservoirs are marked with distinct symbols. Direction of pumps is indicated by an arrow.
- `link_attribute` and `node_attribute` string options are expanded to include attributes for all assets (pipe, valve, and pump for link and junction, tank, and reservoir for node)

Note that relying on `geopandas` for plotting will require elevating `geopandas` from an optional dependency to a required one.

This PR addresses #223 since geopandas plots the vertices, which means #390 is no longer necessary.

## Tests and documentation
Current tests and documentation already cover the `plot_network` function.
An additional test in `test_graphics` is added to cover additional test cases. For development purposes, the old `plot_network` is included in this test file under the name `plot_network_nx`. In `test_plot_network_options` there is a variable `compare` which can be manually set to True to produce plots that side-by-side compare the old and new plotting functions. However, since we don't want to continue support the networkx plotting this isn't ran as part of the default test suite.

## Acknowledgement
By contributing to this software project, I acknowledge that I have reviewed the [software quality assurance guidelines](https://wntr.readthedocs.io/en/stable/developers.html) and that my contributions are submitted under the [Revised BSD License](https://wntr.readthedocs.io/en/stable/license.html).

kbonney committed:
first draft of plot_network function with a GIS backend

kbonney committed:
incorporating the rest of the keywords from plot_network in plot_network_gis

kbonney committed:
Merge branch 'main' into gis_plotting

kbonney committed:
combining link-like and node-like gis files

kbonney committed:
adding tank/reservoir marker shapes and implementating other type options for attributes

kbonney committed:
set geopandas backend to the plot_network function name. update the geopandas backend to handle lists for attributes and add shapes for valves and pumps.

kbonney committed:
archive networkx plotting function in test suite

kbonney committed:
clean up function and add directed functionality

kbonney committed:
add comparison test for plotting

kbonney committed:
handle cbar manually to avoid error in earthquake demo

kbonney commented:
Pending items to be resolved here or saved for later updates:

- Customize asset markers (valve, pump, reservoir, tank) to match the shapes seen in EPANET. Could borrow code from viswaternet.
- Current approach to showing pump direction as well as pipe direction when `directed=True` is sensitive to aspect ratio. It looks correct when the aspect ratio is 1, but if you stretch the image in the pop-up viewer the angles of the arrow directions are not updated, so they look off. Another approach could be taken using [matplotlib arrow objects](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.arrow.html) to avoid this.

kbonney committed:
Merge remote-tracking branch 'usepa/main' into gis_plotting

kbonney committed:
extend test cases

kbonney commented:
Hi @meghnathomas, it seems I cannot directly add you as a reviewer to the PR. This probably has to do with some repository level restriction on who can be added as a reviewer. I'll see what I can do to get you added, but let me know if it causes any problems with the review in the meantime.

kbonney commented:
Failing tests should be addressed when #458 is pulled in.

angusmcb commented:
Will this implementation keep geopandas as an optional dependency, only needed if plotting ? Would be much appreciated if this is the case :-)

meghnathomas commented:
Hi, sorry for the delay on the review! Here's some thoughts/comments:

1. For me, often (but not always!) the color bars seem to be normalized to 1 when I use the new geopandas code.  E.g., these are the plots I get when I run `test_plot_network2` in _test_graphics.py_ (plotting directed and undirected plots of node elevation and link length) 

geopandas plotting undirected             |  networkx plotting undirected
:-------------------------:|:-------------------------:
![image](https://github.com/user-attachments/assets/44c5f963-9024-48f6-8a56-e3e36b55485b) |  ![image](https://github.com/user-attachments/assets/cf23ba3f-1e52-495a-8842-9ccf8c837839)

geopandas plotting directed             |  networkx plotting directed
:-------------------------:|:-------------------------:
![image](https://github.com/user-attachments/assets/edf3a5b2-680b-4e68-b92c-83836fd4c07a) | ![image](https://github.com/user-attachments/assets/9075a992-5a9e-46ea-b91f-2b8d2053ae89)

Similarly, these are the results from `test_plot_network4` :

geopandas            |  networkx 
:-------------------------:|:-------------------------:
![image](https://github.com/user-attachments/assets/224d5755-17b5-4c84-b43f-c8e61c7e669c)  | ![image](https://github.com/user-attachments/assets/acd756f1-06d9-4f1c-9e22-2057ddb25de7)

But when we run` test_plot_network5`, the plots match. The major difference seems to be that we specify  "node_range" when we call the function here. 

geopandas            |  networkx 
:-------------------------:|:-------------------------:
![image](https://github.com/user-attachments/assets/6b250631-b31c-46f9-9728-e1238d082036)  | ![image](https://github.com/user-attachments/assets/ebc81522-be61-4622-a03d-148207b6eaa9)

Am I the only one seeing these discrepancies? These are the package version I'm working with on my machine:
matplotlib 3.7.1
geopandas 0.14.4
numpy 1.26.4
pandas 2.2.3
networkx 3.3

2. It might be useful to add a legend (that you can choose to show or hide) for the pumps (triangles? arrows?), valves (a line?), tanks (diamonds?), and reservoirs (squares?) so they can be differentiated easily!

3. The directed graphs are a lot more visually pleasing in the geopandas version. They might be a little easier to visualize if the arrows were slightly smaller, or could somehow be scaled to the fig size. Not a huge deal at all, just noting for future possible enhancements. (looks like you mentioned this in your first comment here on this thread)

4. In the third set of plots above (when we use network Net1), in the geopandas plot the color bars seem to squeeze the network layout a little bit but the networkx version manages to fit the color bars and plots together quite nicely

5. It also looks like for the networkx plots, wntr automatically labels the color bars "Node" or "Link" (which is a nice touch) but that doesn't seem to happen for the geopandas plots

6. It's very cool that you guys plot vertices too now. Looks great!
![image](https://github.com/user-attachments/assets/79f64a1d-4325-4993-8808-cfabd3d45cd4)

7. Is there any way to disable valve/pump/reservoir/tank plotting? Like for this network, [BWFLNet](https://data.mendeley.com/datasets/srt4vr5k38/1), there are about 500 throttle control valves in the network and the graphic is a little hard to look at: 
![image](https://github.com/user-attachments/assets/a2d79b37-af1c-45e2-9a94-fc13ec39ae34)

8. Do you also plan to show valve direction (like you show pump direction)?

kbonney commented:
> Hi, sorry for the delay on the review! Here's some thoughts/comments:
> 
> 1. For me, often (but not always!) the color bars seem to be normalized to 1 when I use the new geopandas code.  E.g., these are the plots I get when I run `test_plot_network2` in _test_graphics.py_ (plotting directed and undirected plots of node elevation and link length)
> 
> geopandas plotting undirected	networkx plotting undirected
> ![image](https://private-user-images.githubusercontent.com/74310847/387867378-44c5f963-9024-48f6-8a56-e3e36b55485b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4NjczNzgtNDRjNWY5NjMtOTAyNC00OGY2LThhNTYtZTNlMzZiNTU0ODViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxN2E2MWY2ZGYyMjZkNzgwMzU4ZTg0YjFmZDI4YTgzZGFmM2JhMTZlODRhN2ZlZDY0ZGUxMzExZWNmZDNhNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mIjtYGuunHWBUJ7NYRNMLoYe_yYYIS2H7v-v4xU2TyI)	![image](https://private-user-images.githubusercontent.com/74310847/387868267-cf23ba3f-1e52-495a-8842-9ccf8c837839.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4NjgyNjctY2YyM2JhM2YtMWU1Mi00OTVhLTg4NDItOWNjZjhjODM3ODM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiNDY0NGY3YTZiNGUyZmRhZjc2MTkxZmM0ZGQ5OTM3YWMxMDVlNjk1Zjg3ZjVjZDAzNDM2Mjk3ODBhYTExNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3XgALQrDjO1yWsJ42glUisJpSP_B7wPlIjDUBVvFWq0)
> geopandas plotting directed	networkx plotting directed
> ![image](https://private-user-images.githubusercontent.com/74310847/387868397-edf3a5b2-680b-4e68-b92c-83836fd4c07a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4NjgzOTctZWRmM2E1YjItNjgwYi00ZTY4LWI5MmMtODM4MzZmZDRjMDdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhZDliY2M2OTQyN2UyYTFhMDRkMDhlOWEzM2NhZTQ5YzE2YTM2NjVkMTgxNzVmMTU1MDUxNTcwODQzYWM5ODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Zpgbn5mN5YxogCiM23eRWwcUw0HKZvechMBCxl1GYqI)	![image](https://private-user-images.githubusercontent.com/74310847/387868414-9075a992-5a9e-46ea-b91f-2b8d2053ae89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4Njg0MTQtOTA3NWE5OTItNWE5ZS00NmVhLWI5MWYtMmI4ZDIwNTNhZTg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjMjYyMTY2ZGEyM2M2ZjhhYmJiMWQ2MDlkNjQxYTk4ZTNiYzQ3YTk2Y2Y0NGQ0MWEzNGNhNDhiMjhiMDA4OTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WVQwQ2vEZet_Ms7ahumMl4c_m2Kto_aL73h5fpFYa7k)
> Similarly, these are the results from `test_plot_network4` :
> 
> geopandas	networkx
> ![image](https://private-user-images.githubusercontent.com/74310847/387870246-224d5755-17b5-4c84-b43f-c8e61c7e669c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4NzAyNDYtMjI0ZDU3NTUtMTdiNS00Yzg0LWI0M2YtYzhlNjFjN2U2NjljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3NzFkOWI1YzBjM2U5NTJiNGQ3NTc3NjhjMzY1YzExZTU3ODgzNGJmNzRmOWE2YmM0NTczZTk3MzdiOTUzYWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6MSW1dHSI6Hhaa5l2NidLbI6ZRFu1ZK8p6i9QFJS0ww)	![image](https://private-user-images.githubusercontent.com/74310847/387870572-acd756f1-06d9-4f1c-9e22-2057ddb25de7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4NzA1NzItYWNkNzU2ZjEtMDZkOS00ZjFjLTllMjItMjA1N2RkYjI1ZGU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEwZDc1OWYxMWQzNDgwNzJkNTFhYjM3ZDU1ODM3N2ZmYWUyZDViYmI0MDJmZDU5ZjE4YWM4MWYxODQwM2I2YTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.k-EoAmbSuVhbuiwaC23H06jrPiLQ0or8CzE1lj5BKkU)
> But when we run` test_plot_network5`, the plots match. The major difference seems to be that we specify "node_range" when we call the function here.
> 
> geopandas	networkx
> ![image](https://private-user-images.githubusercontent.com/74310847/387870871-6b250631-b31c-46f9-9728-e1238d082036.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4NzA4NzEtNmIyNTA2MzEtYjMxYy00NmY5LTk3MjgtZTEyMzhkMDgyMDM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3ZDllODA0ODVkYTZjNzE5YWI1ZTJjYWE5NjFkOTQzOWZmNGQ2MzRkNzExNmQyMzdjODhmZjZkNjk2MzlhNzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.j3qEUY9-NgeK7g84qZDc36ZCKbWAhA3U9BnTa1TA1Cc)	![image](https://private-user-images.githubusercontent.com/74310847/387871596-ebc81522-be61-4622-a03d-148207b6eaa9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4NzE1OTYtZWJjODE1MjItYmU2MS00NjIyLWEwM2QtMTQ4MjA3YjZlYWE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxMGRmNjVjMTU2NDJhNWU4ZGM4NmVmMmYzMmU4YTdhOGViYTI1NDY2NTYxZTljM2Q2NGUxMTM3OWY3YjNiMmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.b8sN8Vm-w4J0s0UfFaXcRBwcaDf4YDgn1XKFNTk_pwo)
> Am I the only one seeing these discrepancies? These are the package version I'm working with on my machine: matplotlib 3.7.1 geopandas 0.14.4 numpy 1.26.4 pandas 2.2.3 networkx 3.3
> 
> 2. It might be useful to add a legend (that you can choose to show or hide) for the pumps (triangles? arrows?), valves (a line?), tanks (diamonds?), and reservoirs (squares?) so they can be differentiated easily!
> 3. The directed graphs are a lot more visually pleasing in the geopandas version. They might be a little easier to visualize if the arrows were slightly smaller, or could somehow be scaled to the fig size. Not a huge deal at all, just noting for future possible enhancements. (looks like you mentioned this in your first comment here on this thread)
> 4. In the third set of plots above (when we use network Net1), in the geopandas plot the color bars seem to squeeze the network layout a little bit but the networkx version manages to fit the color bars and plots together quite nicely
> 5. It also looks like for the networkx plots, wntr automatically labels the color bars "Node" or "Link" (which is a nice touch) but that doesn't seem to happen for the geopandas plots
> 6. It's very cool that you guys plot vertices too now. Looks great!
>    ![image](https://private-user-images.githubusercontent.com/74310847/387879860-79f64a1d-4325-4993-8808-cfabd3d45cd4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4Nzk4NjAtNzlmNjRhMWQtNDMyNS00OTkzLTg4MDgtY2ZhYmQzZDQ1Y2Q0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjZjlkMWYyZjc4NWM2Y2I5ZWZjZDViYjY0NjRkYzc1MzNhYTJiOTM3NWFhNmNhZDg0MzU0ZGRlY2ZiMWJjYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._DwFdXdpYVeJhwPcGW3FOBprqxgffoe15uUEC-8s8hg)
> 7. Is there any way to disable valve/pump/reservoir/tank plotting? Like for this network, [BWFLNet](https://data.mendeley.com/datasets/srt4vr5k38/1), there are about 500 throttle control valves in the network and the graphic is a little hard to look at:
>    ![image](https://private-user-images.githubusercontent.com/74310847/387880198-a2d79b37-af1c-45e2-9a94-fc13ec39ae34.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNjI1ODUsIm5iZiI6MTczMjA2MjI4NSwicGF0aCI6Ii83NDMxMDg0Ny8zODc4ODAxOTgtYTJkNzliMzctYWYxYy00NWUyLTlhOTQtZmMxM2VjMzlhZTM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDAwMjQ0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkNTY4MTM3ZDhkYmE5ZmZiZjYyM2Y5MTNjZTQ3NjY4MmM1YWM0NDk4MDhkYmU1ODk1MjllNDgxNWU5YWYxYWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iXWAOHkWTK6nBPV1IjudBn6-dzkkTUzqjWYVtT9GbqM)
> 8. Do you also plan to show valve direction (like you show pump direction)?

Hi @meghnathomas, thanks for your rigorous testing of the PR. I'll work through the items you pointed out and get back to you with some updates .

kbonney committed:
Merge remote-tracking branch 'usepa/main' into gis_plotting

kbonney committed:
Merge remote-tracking branch 'usepa/main' into gis_plotting

kbonney committed:
fix bug caused by node_type no longer provided by GIS geodataframes, fix bugs pointed out by meghna: colorbar scale and spacing, reduce directed arrow size, add option for pumps/valves (default to not plot),

kbonney committed:
add extra test cases, remove unneccesary calls to plt.figure

kbonney committed:
setting compare to false

kbonney commented:
@meghnathomas,
1. Good catch, this should be fixed by the latest commits.
2. I agree this would be helpful. I will have to spend more time looking into this.
3. I agree that the arrows were too large. I changed them so they are more like the size in the networkx version.
4. The cbars should have the correct spacing now.
5. This originally was an intentional choice by the team. I think it looks okay when only plotting one colorbar, but it is confusing if you have both. I'll revisit this with the others.
6. Yes! This PR originated as an effort to add vertices in the networkx framework. It became a pain so it started to make more sense to use the geopandas plotting capabilities, which are much more flexible and user-friendly.
7.  I added additional keyword args for this. They default to not show pumps and valves so that previous plotting code doesn't retroactively add them.
8. This should be simple to add. I will look into it.

I also caught a bug where labels only rendered for junctions/pipes rather than all assets.

meghnathomas commented:
@kbonney awesome, thanks, everything looks good with the colorbar values, spacing b/w plot and colorbar, arrow sizes, and suppressing pump/valve plotting!

A couple of comments:
1. I think in the networkx version, when alpha < 1, the color bar alpha would match the marker colors alpha, but i think in this geopandas version alpha only applies to the markers and not to the color bar
2. Are the tanks, reservoirs, pumps, and valves supposed to show up by default on all plots, even when you specify node_attribute/link_attribute? Or are they only supposed to show up on a basic network layout plot with no data assigned to nodes/links? If a user specifies "node_attribute" and the list of nodes they are plotting values for doesn't contain the tanks and reservoirs, those nodes don't show up on the map so I don't know where the tanks/reservoirs are located. Is that an intentional choice? If so, that's fine, just wanted to check! Noticed this when I used the following kwargs in your new _test_graphics_ file:
```
{"node_attribute": node_list},
{"node_attribute": random_node_values},
{"node_attribute": random_node_dict_subset}
```

I didn't look closely enough at the figures to check if this is true for pumps (and valves) too, but this question applies to them as well

kbonney committed:
add alpha to colorbars

kbonney commented:
Thanks for the additional comments @meghnathomas:
1. Fixed this, the alpha should now carry over to the cbars.
2. All nodes are always plotted, unless a subset of nodes is specified in which case only that subset is plotted. This includes junctions, tanks, and reservoirs. The only difference is that tanks/reservoirs have a different shape and size.
This should also be the case for links (pipes, pumps, valves). The edge of a pump or valve is always plotted by default, but the additional visual indicators are only shown when `plot_pumps` or `plot_valves` is set to True.

kbonney committed:
fix color bug on mpl 3.8, change plot_valve/pumps kwarg name, set aspect to "equal"

kbonney committed:
fix test case to adjust for new kwarg names

kbonney committed:
add legend and plot all nodes regardless of node_attribute

kbonney committed:
add additional tests

kbonney committed:
remove old plot network code, clean up plot network, add default colorbar label when passing a str for attribute, fix missingkwds markersize

kbonney committed:
remove pump/valve direction plotting tests

kbonney committed:
update plot_network to use _format functions instead of _prepare function

kbonney commented:
Summary of recent updates:
- Removed pump/valve direction code. This feature could be added back in a later update, but for now I want to prioritize getting a working version of the new function that is backwards compatible.
- When a string is provided as an attribute it is used as the default label.
- Added a legend. The legend chooses some color from the plotted elements as a representative for the symbol. Ideally we would set the color to always be black for consistently, but overriding this behavior has proven difficult.
- All nodes are now plotted regardless of attribute. If a node is not included by the provided attribute the color defaults to black. Similar behavior for links.

meghnathomas commented:
@kbonney this looks awesome and works super well! :-) Sorry to continue to pester you with comments, but I have just a couple more based on the most recent version of the code:
1. It looks like the colorbar titles can sometimes overlap a little - is that because the aspect ratio is now "auto" you think?
![image](https://github.com/user-attachments/assets/c5e466af-f163-407d-ab95-5b33dfff0ddd)
2.  Thank you for adding the legend! Please feel free to ignore this comment or leave this task for another time - but it might also be useful to allow users to customize legend location so they can move it around the plot if they want. Providing a default nice legend location that's always out of the way of the network plot no matter what the shape of the network is is also an option but a lot more work on your end probably (that's why we didn't do that for viswaternet haha). In any case this is not critical at all!

angusmcb commented:
I've been working on a plugin for QGIS (currently experimental) which effectively replaces the wntr gis module and geopandas with PyQGIS: 
https://github.com/angusmcb/wntr-qgis/

I did think that Geopandas wasn't included as a dependency as standard with QGIS, but have now realised that it _is_ included with the most recent QGIS (windows) releases, so no problem!

kbonney commented:
> I've been working on a plugin for QGIS (currently experimental) which effectively replaces the wntr gis module and geopandas with PyQGIS: https://github.com/angusmcb/wntr-qgis/
> 
> I did think that Geopandas wasn't included as a dependency as standard with QGIS, but have now realised that it _is_ included with the most recent QGIS (windows) releases, so no problem!

Thanks for the clarification. It is exciting to hear about the QGIS plugin - looking forward to its progress!

dbhart reviewed (approved):
This PR makes a huge improvement in the WNTR visualization, and I believe that, while there are always more features to add, that this effectively solves the original problem and should go ahead and be pulled into main.

kbonney committed:
add geopandas backend as an option to plot_network. defaults to networkx backend.

kbonney committed:
add backend to docstring

kbonney committed:
add ax returning and information on gpd option.
